### PR TITLE
External pass

### PR DESCRIPTION
**Pull request by Danacus**
This PR implements support for the external pass API, enabling the creation of MLIR passes in Rust. It's still a bit rough on the edges, this is an early draft.

The main idea is as follows:
- Users of `melior` create a struct and implement the `ExternalPass` trait.
- When calling `create_external` on this struct, we use the external pass API to create a pass on the MLIR side.
  - Using generic functions, callback functions are generated for this struct and passed to MLIR.
  - Ownership of the Rust pass struct is conceptually moved to MLIR as a pointer to this struct is passed as `userData`, hence `Box::into_raw` is used.
- When the callbacks are called, we cast the `userData` pointer to the appropriate Rust struct and call a method of the `ExternalPass` trait.
- When MLIR decides to destruct the C++ counterpart of this pass, we make sure to clean up our struct (manually `drop_in_place`), since it won't be freed by MLIR, and ownership is no longer tracked by Rust after leaking to `Box`.

The following code snippet illustrates how this can be used:

```rust
#[repr(align(8))]
struct TestType(u8);

static TEST_TYPE: TestType = TestType(0);

#[derive(Clone, Debug)]
struct TestPass {
    value: i32,
}

impl<'a> ExternalPass<'a> for TestPass {
    fn name(&self) -> StringRef<'a> {
        StringRef::from("test pass")
    }

    fn argument(&self) -> StringRef<'a> {
        StringRef::from("test argument")
    }

    fn description(&self) -> StringRef<'a> {
        StringRef::from("a test pass")
    }

    fn op_name(&self) -> StringRef<'a> {
        StringRef::from("")
    }

    fn dependent_dialects(&self) -> Vec<melior::dialect::DialectHandle> {
        vec![]
    }

    fn type_id(&self) -> TypeId {
        TypeId::create(&TEST_TYPE)
    }

    fn construct(&mut self) {
        println!("Constructed pass!");
    }

    fn initialize(&mut self, context: ContextRef<'_>) {
        println!("Initialize called!");
    }

    fn run(&mut self, operation: OperationRef<'_, '_>) {
        println!("Running test pass!");
        operation.dump();
    }
}

fn main() {
    let registry = DialectRegistry::new();
    register_all_dialects(&registry);

    let context = Context::new();
    context.append_dialect_registry(&registry);
    context.load_all_available_dialects();

    let location = Location::unknown(&context);
    let mut module = Module::new(location);

    let index_type = Type::index(&context);

    module.body().append_operation(func::func(
        &context,
        StringAttribute::new(&context, "add"),
        TypeAttribute::new(
            FunctionType::new(&context, &[index_type, index_type], &[index_type]).into(),
        ),
        {
            let block = Block::new(&[(index_type, location), (index_type, location)]);

            let sum = block.append_operation(arith::addi(
                block.argument(0).unwrap().into(),
                block.argument(1).unwrap().into(),
                location,
            ));

            block.append_operation(func::r#return(&[sum.result(0).unwrap().into()], location));

            let region = Region::new();
            region.append_block(block);
            region
        },
        &[],
        location,
    ));

    let pass_manager = PassManager::new(&context);

    let test_pass = TestPass { value: 10 };
    pass_manager.add_pass(test_pass.create_external());
    pass_manager.run(&mut module).unwrap();

}

```

Things that still need to be done:

- [ ] Write tests
- [ ] Write documentation with example of usage
- [ ] Make it a bit more ergonomic to use (e.g. the `TypeId` is still a bit awkward, support for `Fn` could be useful)

### CODE REVIEW

**Comment by Danacus** [comment]
Update (https://github.com/raviqqe/melior/pull/251/commits/99f8574755bfa638b6fe3f259bbc6ff480b66683): using `mlirTypeIDAllocatorAllocateTypeID` to avoid needing to define one manually. As a consequence, each instance of a certain pass has a unique type ID. This might not be the best solution, but this way we no longer need to define `ExternalPass::type_id`.

**Comment by Danacus** [comment]
I've simplified the `ExternalPass` trait and moved `create_external` out of the trait. This way it's possible to implement `ExternalPass` for `Fn(OperationRef<'c, '_>)` (not sure if `FnMut` or `FnOnce` should be allowed). Then `ExternalPass` could be used as follows:

```rust
impl<'c> ExternalPass<'c> for TestPass {
    fn construct(&mut self) {
        println!("Constructed pass!");
    }

    fn initialize(&mut self, context: ContextRef<'c>) {
        println!("Initialize called!");
    }

    fn run(&mut self, operation: OperationRef<'c, '_>) {
        println!("Running test pass!");
        operation.dump();
    }
}

impl TestPass {
    fn create(self) -> Pass {
        create_external(self, "test pass", "test argument", "a test pass", "", &[])
    }
}

// [...]

// Create using `TestPass::create()`
pass_manager.add_pass(test_pass.create());

// Create from `Fn` closure
pass_manager.add_pass(create_external(
     |operation: OperationRef| {
         operation.dump();
     },
     "test closure",
     "test argument",
     "test",
     "",
     &[],
));
pass_manager.run(&mut module).unwrap();
```

Please let me know if you prefer the previous implementation over this one, since I'm not sure which one I prefer either.

**Comment by Danacus** [comment]
When trying to use this code I noticed that since the callbacks take an `OperationRef`, passes cannot mutate operations. We would need a mutable reference to an `Operation` without having an owned `Operation` (the operation is "owned" by MLIR, so we don't want to construct an owned `Operation`). Creating a `OperationMut` similar to `OperationRef`, but that implements `DerefMut` would solve this, but I'm not sure if this is the best solution.

**Comment by raviqqe** [comment]
> Update ([99f8574](https://github.com/raviqqe/melior/commit/99f8574755bfa638b6fe3f259bbc6ff480b66683)): using `mlirTypeIDAllocatorAllocateTypeID` to avoid needing to define one manually. As a consequence, each instance of a certain pass has a unique type ID. This might not be the best solution, but this way we no longer need to define `ExternalPass::type_id`.

We should probably use the old API because the documentation of the C API says that generated type IDs are only valid for the lifetime of the allocator.

I'm going to change the type ID type into TypeId<'c> after we merge these changes. So you can use the TypeId type as it is in this PR.

File: melior/src/pass/external.rs [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+            StringRef::from(argument).to_raw(),
+            StringRef::from(description).to_raw(),
+            StringRef::from(op_name).to_raw(),
+            dep_dialects_raw.len() as isize,
+            dep_dialects_raw.as_mut_ptr(),
+            callbacks,
+            pass_box as *mut std::ffi::c_void,
+        );
+        Pass::from_raw(raw_pass)
+    }
```
Comment (raviqqe) [inline melior/src/pass/external.rs]: I think we can pass in the pass ID (of `TypeId`) directly to this function for now to keep the API similar to the C API and not to deal with the `TypeId` creation issue. In this way, we can still keep the `impl<F: Fn> ExternalPass for F`.

**Review by raviqqe — COMMENTED** [review]
Thank you for the changes!

**Comment by raviqqe** [comment]
> When trying to use this code I noticed that since the callbacks take an `OperationRef`, passes cannot mutate operations. We would need a mutable reference to an `Operation` without having an owned `Operation` (the operation is "owned" by MLIR, so we don't want to construct an owned `Operation`). Creating a `OperationMut` similar to `OperationRef`, but that implements `DerefMut` would solve this, but I'm not sure if this is the best solution.

I think it's a good way forward. I can make those changes after this PR is merged. You can just keep using `OperationRef` here.

**Comment by raviqqe** [comment]
> I've simplified the `ExternalPass` trait and moved `create_external` out of the trait. This way it's possible to implement `ExternalPass` for `Fn(OperationRef<'c, '_>)` (not sure if `FnMut` or `FnOnce` should be allowed). Then `ExternalPass` could be used as follows:
> 
> ```rust
> impl<'c> ExternalPass<'c> for TestPass {
>     fn construct(&mut self) {
>         println!("Constructed pass!");
>     }
> 
>     fn initialize(&mut self, context: ContextRef<'c>) {
>         println!("Initialize called!");
>     }
> 
>     fn run(&mut self, operation: OperationRef<'c, '_>) {
>         println!("Running test pass!");
>         operation.dump();
>     }
> }
> 
> impl TestPass {
>     fn create(self) -> Pass {
>         create_external(self, "test pass", "test argument", "a test pass", "", &[])
>     }
> }
> 
> // [...]
> 
> // Create using `TestPass::create()`
> pass_manager.add_pass(test_pass.create());
> 
> // Create from `Fn` closure
> pass_manager.add_pass(create_external(
>      |operation: OperationRef| {
>          operation.dump();
>      },
>      "test closure",
>      "test argument",
>      "test",
>      "",
>      &[],
> ));
> pass_manager.run(&mut module).unwrap();
> ```
> 
> Please let me know if you prefer the previous implementation over this one, since I'm not sure which one I prefer either.

Because the MLIR C API provides the `initialize` callback, it can't be `FnOnce` but probably `FnMut` is fine.

**Comment by raviqqe** [comment]
Sorry my comments are a bit scatttered around. In summary,

- You can keep the current implementation of `TypeId` and `OperationRef` that are not incompatible with this PR somewhat. I can make proper changes to their API after this PR is merged.
- We can implement `ExternalPass` for `FnMut`. You can make changes to this PR if you want to.
- We can keep the original implementation of `TypeId::create` (the one with `mlirTypeIDCreate`).
  - The one using `mlirTypeIDAllocator` seems to be not valid according to the C API documentation.
- `create_external` should accept `pass_id: TypeId` instead of generating one automatically as we have several issues by doing that.
  - When we use `mlirTypeIdCreate`, we need to make sure the pointer is always 8-byte aligned.
  - We deviate the API from the original C API.

Let me know if you have any opposition to this direction!

**Comment by Danacus** [comment]
Thank you for your comments! I will make the suggested changes, and I'll make clippy happy such that the checks succeed.

**Comment by raviqqe** [comment]
> - [ ] Write tests
> - [ ] Write documentation with example of usage

Are you gonna work on those in this PR? I can make those changes if you want me to.

**Review by raviqqe — APPROVED** [review]

**Comment by Danacus** [comment]
I'm adding some documentation right now, which I will push in a few minutes. I won't have time to write tests until later today, so feel free to write them yourself if you like. Otherwise, I'll write them later today.

**Comment by raviqqe** [comment]
OK. Then, I'll just wait for you to implement them. Thanks a lot for the work!

**Comment by Danacus** [comment]
Something I noticed while writing tests is that there doesn't seem to be a `ModuleRef` yet, so `Module`-specific methods cannot be used on a `OperationRef`. Perhaps `ModuleRef` (and `ModuleMut`) would be useful additions?

**Comment by raviqqe** [comment]
Modules are merely a built-in top-level `module` operation. So at least, modification to it is accessible even now. Which specific methods of `Module` do you want to use?

**Comment by raviqqe** [comment]
Thank you for the changes!

File: melior/src/pass/external.rs [hunk 2]
```diff
@@ -226,6 +226,17 @@ mod tests {
...
                 self.value = 30;
                 assert!(operation.verify());
+                assert!(
+                    operation
+                        .region(0)
+                        .expect("module has a body")
+                        .first_block()
+                        .expect("module has a body")
+                        .first_operation()
+                        .expect("body has a function")
```
Comment (Danacus) [inline melior/src/pass/external.rs:236]: This code could be replaced with `Module::body` if `OperationRef` was a `ModuleRef`. It's not that important, just a convenience.

**Comment by Danacus** [comment]
Thank you for merging this PR!

**Pull request merged**